orderBook.js

```diff
@@ -1,5 +1,59 @@
+/* eslint-disable indent */
 function reconcileOrder(existingBook, incomingOrder) {
-  throw new Error('Not Implemented Yet')
+
+    var updatedBook = []
+
+
+    if (existingBook.length == 0) {
+
+        updatedBook.push(incomingOrder)
+        return updatedBook
+    }
+    else if (existingBook[0]['type'] == 'sell' && incomingOrder['type'] == 'sell') {
+
+
+        updatedBook.push(existingBook[0], incomingOrder)
+
+    }
+    else if (existingBook[0]['type'] == 'buy' && incomingOrder['type'] == 'sell') {
+        if (existingBook[0]['quantity'] == incomingOrder['quantity'] && existingBook[0]['price'] == incomingOrder['price']) {
+
+            return updatedBook
+
+        }
+        else if (existingBook[0]['quantity'] != incomingOrder['quantity'] && existingBook[0]['price'] != incomingOrder['price']) {
+
+            updatedBook.push(existingBook[0], incomingOrder)
+
+        }
+        else if (existingBook[0]['quantity'] == incomingOrder['quantity'] && existingBook[0]['price'] != incomingOrder['price']) {
+            if (existingBook[0]['price'] < incomingOrder['price']) {
+
+                updatedBook.push(existingBook[0], incomingOrder)
+
+            }
+            else if (existingBook[0]['price'] > incomingOrder['price']) {
+
+                return updatedBook
+            }
+
+        }
+        else if (existingBook[0]['quantity'] > incomingOrder['quantity'] && existingBook[0]['price'] == incomingOrder['price']) {
+
+            updatedBook.push(existingBook[0])
+            updatedBook[0]['quantity'] = existingBook[0]['quantity'] - incomingOrder['quantity']
+
+        }
+        else if (existingBook[0]['quantity'] < incomingOrder['quantity'] && existingBook[0]['price'] == incomingOrder['price']) {
+
+            updatedBook.push(incomingOrder)
+            updatedBook[0]['quantity'] = incomingOrder['quantity'] - existingBook[0]['quantity']
+
+        }
+    }
+
+    return updatedBook
+
 }
```

Comment:
This much logic and branches in a single function should never happen. This should be many sub functions.

Also you are only looking at the first order in the existing book. What happens if there are three orders in the existing book and the 3rd one would have matched. Your code would never fulfill that order even though it should. In fact it's possible that your code would never fulfill any orders because it's only looking at the first one in the book.